lib/Driver/Driver.cpp

```diff
@@ -986,9 +1001,11 @@ void Driver::buildInputs(const ToolChain &TC,
           Ty = types::TY_Object;
         }
       }
-
-      if (checkInputExistence(*this, Args, Diags, Value))
-        Inputs.push_back(std::make_pair(Ty, A));
+      if (checkInputExistence(*this, Args, Diags, Value)) {
+        // Substitute the OPT_index_file_path Arg if this is one of them.
+        const Arg *IA = IndexFilePathArgs.lookup(Value);
+        Inputs.push_back(std::make_pair(Ty, IA ? IA : A));
```

Comment:
New round of comments where I've given up on improving `-index-file`: I think this is too clever. Yes, it gets the information where you want it to go, but at the cost of making `-index-file-path` "input-like", which it really wasn't before. I'd rather just go back to the list of arguments and see if the name matches up later.

Reply:
I have pushed a version with index file path handling that I hope is better.